### ISSUE TITLE
aria2 not downloading torrent/magnet. File URL works

### ISSUE THREAD

**Issue by jabardast** [aria2/aria2#679]
I am running a minimal image of pi.

H/W Details-
RASPBERRY PI MODEL B 756-8308 Raspberry Pi B
TP-LINK TL-WN725N Wireless N Nano USB Adapter
Uname -a output --> Linux pi 3.12.25+ #700 PREEMPT Thu Jul 24 17:51:46 BST 2014 armv6l GNU/Linux

I was using pi as a headless download server using aria2 (aria2 version 1.15.1).
pi was connected to internet via Wifi with my ISP provided wifi router. LAN port of PI was not used at all.
All was working great up until 2 days back.

2 Days ago, i switched to a different ISP with a better deal. To save money, I declined to buy their wifi router.
They installed a CAT5 cable and gave a static IP. I tested it with my laptop, all worked well.

After 2 days of googling, I managed to configure Wifi Access point on my pi.
Now my external IP is configured on pi Ethernet card. And wifi card provided access point to my devices.
I am using hostapd and udhcpd.

Today I noticed that aria2 running on my pi (I access it via webgui) is not downloading any torrents or magnet link.
If I give a direct URL (like - http://dl2.filehippo.com/811e3a59d4db46 ... 171c14d5b0) via webgui,
it works. But torrents doesn't work.

Debug-log file is attached. Only relevant error i can see was

Exception: [CreateRequestCommand.cc:100] errorCode=-1 No URI available.
2016-06-13 07:50:01.401733 DEBUG - [AbstractCommand.cc:426]CUID#24 - Aborting download
2016-06-13 07:50:01.401893 DEBUG - [AbstractCommand.cc:376]CUID#24 - Not trying next request. No reserved/pooled request is remaining and total length is still unknown.

I thought may be my ISP is blocking torrent traffic.
So I tried downloading the same torrent on my laptop via deluge-portable.
It's working fine and with good speed too.

Does aria2 configuration needs to be tweaked with the new setup ?

Can anyone one point me to a possible fix for this issue ?

Thanks in advance.

Regards,
Jab.
[downloads.zip](https://github.com/aria2/aria2/files/313249/downloads.zip)

**Comment by tatsuhiro-t** [comment]
Make sure that port forwarding is done for both UDP and TCP to aria2 through firewall since aria2 does not have capability such as upnp.
The above log suggests that no tracker URI is found.  If torrent only contains udp tracker, you have to wait for DHT to look up for peers.  udp tracker support has been added since v1.20.0.

**Comment by jabardast** [comment]
thanks for the tip sir. I will see what I can do.
May be it will be easier to upgrade aria2. Right now apt-get on pi only installs 1.15.1 version.
I appreciate your help. I will come back with the results

**Comment by jabardast** [comment]
After a failed attempt to compile latest aria2 , I started looking into routing of pi.
Apparently, my routing table needed few more entries.

While setting up Wifi Router + Access point - I had enabled NAT via
 iptables –t nat –A POSTROUTING –o eth0 -j MASQUERADE

I needed following entries too
Tell it to forward all traffic to the Internet:
iptables -A FORWARD -i wlan0 -o eth0 -j ACCEPT

Finally, tell it to forward returning Internet data to the appropriate client:
iptables -A FORWARD -i eth0 -o wlan0 -m state --state RELATED,ESTABLISHED -j ACCEPT

I also disabled rpc-listen-all=true in my conf file
 Now aria2 is listening on my internal IP address not on External IP.

Thanks for the Author of this website - http://makezine.com/projects/browse-anonymously-with-a-diy-raspberry-pi-vpntor-router/

Thanks to tatsuhiro-t  for the pointer.
